updated liveness and readiness probe descriptions (#106)
adoptingerlang/adoptingerlang: adoptingerlang.org

Copying files into an image and caching layers are no longer the only options for efficiency when building a Docker image. Caching the built dependencies in a layer is all well and good but that layer also contains the Hex package cache Rebar3 creates under =~/.cache/rebar3/hex=. Any change to =rebar.config= or =rebar.lock= will result in all the packages having to not only be rebuilt but also be re-fetched from Hex. Additionally, the instruction to copy in the whole project, creating an additional layer with all the source is wasteful since we only care about the build artifacts.

These issues are resolved as of Docker 19.03 through an experimental syntax for
mounting files to the context of a =RUN= command. To enable the experimental
syntax the environment variable =DOCKER_BUILDKIT= must be set or
={"features":{"buildkit": true}}= be set in =/etc/docker/daemon.json=, and =#

> syntax=docker/dockerfile:1.2= used as the first line of the Dockerfile:

Reviewer: is splitting the markup really working here where the `=#` is on one line and the rest on another one?

Author: Yup, I just checked the preview deploy https://651400cdecfab7009b23a4fb--inspiring-shannon-000ae0.netlify.app/docs/production/docker/

whether it is during a deploy, =livenessProbe= failure or scale down. Erlang will
call =init:stop()= when =SIGTERM= is sent by Kubernetes to signal shutdown. As
discussed in [[Releases]], each Applications will be stopped in the reverse they
were started and each Application's supervision tree terminates children in
reverse order.

By default the process can wait up to 30 seconds (configurable with the
=terminationGracePeriodSeconds= =PodSpec= option) before Kubernetes sends a =SIGKILL=
signal that will force the process to terminate.

An important note about deployments and =readinessProbe= is the behaviour if a
=readinessProbe= never passes. During deployment if the =Pod= never passes its
=readinessProbe= then, depending on the deployment strategy, =Pods= from the
previous version of the deployment that are =Ready= may remain until they do. This
is covered more in the follow section [[Rolling Deployments]].

In the =service_discovery= project only =readinessProbe= is defined for the purpose
of removing the =Pod= from =Service= backends without killing the container:

#+BEGIN_SRC yaml
readinessProbe:
  httpGet:

>     path: /ready

Reviewer: might be interesting to eventually aim for `/health/<type>` but we're good to go here.

The volume =migrations= exists only in memory and is used to copy the current SQL migration files from the release image to a shared directory that the =flyway= container will use to run validation. The =emptyDir= type of volume with =medium= =memory= is used because we are not using this volume to persist any data, only to share between two =Init Containers=. If =medium: Memory= was not included, the =emptyDir= is created on the host file system but it is still empty initially for each new =Pod=, and is removed when the =Pod= is deleted.

**** Readiness, Liveness and Startup Probes

> Each container in a =Pod= can define a =readiness=, =liveness= and =startup= probe. The

Reviewer: ```suggestion
Each container in a =Pod= can define a =readiness=, a =liveness=, and a =startup= probe. The
```
welcome to oxford commas

=liveness= probe . A startup probe can be added to delay when the =liveness= and

> =readiness= probes start being checked.

Reviewer: ```suggestion
=readiness= probes start being checked when the pod starts.
```

Probes are defined either as a command to execute in the container, as an HTTP
=GET= request made to a given port and path, or as an attempt to open a TCP
connection to a specific port.

> A =livenessProbe= is not required to ensure restart of failed containers. If the
> process in the container crashes or stops for any reason, it will already be
> restarted. The probe should only be used for cases the process does not crash
> but for some reason has deadlocked and can not respond to requests. For this

Reviewer: ```suggestion
While kubernetes will restart any failed container, there are some times where a
process becomes unresponsive, such as deadlocks. A =livenessProbe= can be
used for such cases, where stopping to respond prompts a restart. For this
```
Trying to state what it is for, rather than what it isn't for first.

reason if a =livenessProbe= is included it should be kept extremely simple. It is
more often than not safer to leave out the =livenessProbe= entirely and rely on
the node crashes, metric based alerting and auto-scaling.

One part of auto-scaling is based on the =readinessProbe=. The =readinessProbe=
tells Kubernetes the =Pod= is ready to receive traffic. When a =Pod= becomes =unready=
because a container is failing the =readinessProbe=, it is removed from the

> =Service= resulting in traffic being directed at the existing =Pods=, increasing

Reviewer: ```suggestion
=Service= resulting in traffic being directed at the remaining =Pods=, increasing
```

their load, then a new =Pod= will be started.

Not using the =livenessProbe= for this purpose means the containers are still
available, since they aren't killed, and the reason for failing the check can be
examined. This may mean attaching a shell if it isn't completely frozen or
forcing a crash dump to be written. In a very resource constrained environment
you may prefer to have the =livenessProbe= in order to have deadlocked containers
and their =Pods= cleaned up right away, instead of waiting for manual intervention

> and the auto-scaler to scale down after load falls.

Reviewer: A tricky pattern here is that if you use readiness probes to help you scale, and your auto-scaler is based on CPU usage, you can find yourself in a situation where during overload you want to refuse traffic, wall yourself off, and start gradually decreasing the resource pool you have, which in turn makes the problem worse.

Author: Do you mean a use-case of failing the readinessProbe during overload?

Reviewer: It might be me who's wrong, but I thought the HPA only cared about readiness on the initial pod boot time, and not after that; so if down the road the pod sets itself as unready when it was and starts using less CPU, then the cluster's CPU average is going down and can prevent actual autoscaling.

Another use for the =readinessProbe= is if the application has to periodically do work
where you don't want to be handling requests, or have to do some sort of
maintenance, it can return a =503= status response for readiness causing it to be

> removed from the =Service= backends.

Reviewer: Specifically here a common use is to allow draining connections while doing remaining work before shutting down.

Author: Na, it isn't used for that. During shutdown the Pod is removed from the Service already. I don't know that it'd hurt, but it isn't necessary.

> Nothing special is needed to have a graceful shutdown, for any reason, whether

Reviewer: Eeeeh, if you don't remove yourself from the load balancer pool, some gateways (like ALBs or ELBs) may still try to send traffic to a process that's halfway through shutting down, and that will cause interrupted connections that are forwarded to the client. This may however be stack dependent because not all stacks clean up as nicely and orderly as an OTP tree.

Author: Removal should already be handled by k8s. There is always delay, even in the case of the readinessProbe being setup to return 503 on shutdown since the check is only done every `period` seconds, but its handled by k8s and not the application itself is my point. Maybe it needs rewording?

Reviewer: Yeah that could make it clearer. It's possible this was just my misunderstanding and I didn't realize that being in a shutting down state also automatically took you out.

Author: Actually, I should probably mention the use of `prep_stop`. I haven't implemented this in `service_discovery` because it is a little tricky in its case since it has 2 applications, one for http and one for grpc, that both have to wait until active requests are handled before shutting down.

I suppose I should also mention the need to have a proper boot order and supervision tree so that stuff like the database pool aren't shutdown before the http app can pause to complete open requests.

I could just add a "see Graceful Shutdown section (to come)" for more...

Author: Eh, maybe it isn't tricky, was overthinking it. It is the same whether they are both in the release or not. Each can wait for their active connections one after the other and the SIGKIll takes care of it if they take too long.

Author: Wait, yea, this is just k8s specific. I will just add a callout to another section and with saying in k8s you don't need to do anything special, but see this section for how to handle the shutdown in your Erlang application.

it is during a rolling deploy or scale down. Erlang will call =init:stop()= when
=SIGTERM= is sent by Kubernetes to signal shutdown, so each individual OTP
Application should perform any necessary shutdown work in their =stop/1= function
of the Application behaviour module as usual. In the case of =service_discovery=
there is no shutdown work aside from trying to complete ongoing requests
successfully. The Applications =service_discovery_http= and =service_discovery_grpc=
can implement logic to delay the return from =stop= until active requests have
been handled.

By default the process can wait up to 30 seconds (configurable with the
=terminationGracePeriodSeconds= =PodSpec= option) before Kubernetes sends a =SIGKILL=
signal that will force the process to terminate.

> An important note about deployments and =readinessProbe= is the behaviour if a
> =readinessProbe= never passes. During a deployment if the container in the =Pod=
> never passes its =readinessProbe= then =Pods= from the previous version of the
> deployment that are =Ready= remain until they do.

Reviewer: ```suggestion
An important note about deployments and =readinessProbe=: 
During a deployment, if the container in the newly deployed =Pod=
never passes its =readinessProbe=, then =Pods= from the previous version of the
deployment that are =Ready= stick around until they do.
```

As a specific note, this is only true if you are using a `spec.strategy.type` of `RollingUpdate` rather than `Recreate`, and two options are in play here as well:

- `maxSurge`: The number of pods that can be created above the desired amount of pods during an update
- `maxUnavailable`: The number of pods that can be unavailable during the update process

K8s will happily kill up until `maxUnavailable` pods during a replacement before it starts hanging on the older ones because the new ones aren't around afaict.

Reviewer: We may want to move that concern to the RollingDeployment section below.

Author: Updated.
